hs_file_types/forms.py

```diff
@@ -0,0 +1,282 @@
+        if executed_by_url:
+            # check that it's an active url
+            url = Request(executed_by_url)
+            try:
+                urlopen(url)
+            except (URLError, HTTPError):
+                self.add_error("executed_by_url", "Provided URL seems like a broken URL")
+        return executed_by_url
+
+    def update_metadata(self, metadata):
+        executed_by = self.cleaned_data['executed_by']
+        executed_by_url = self.cleaned_data['executed_by_url']
+        if executed_by:
+            metadata.executed_by = executed_by
+            metadata.executed_by_url = None
+        elif executed_by_url:
+            metadata.executed_by = None
+            metadata.executed_by_url = executed_by_url
+        else:
+            metadata.executed_by = None
+            metadata.executed_by_url = None
+
+        metadata.has_model_output = self.cleaned_data['has_model_output']
+        metadata.is_dirty = True
+        metadata.save()
+
+
+class ModelProgramMetadataValidationForm(forms.Form):
+    version = forms.CharField(required=False, max_length=250)
+    release_date = forms.DateField(required=False)
+    website = forms.URLField(required=False, max_length=255)
+    code_repository = forms.URLField(required=False, max_length=255)
+    programming_languages = forms.CharField(required=False)
+    operating_systems = forms.CharField(required=False)
+    mi_json_schema = forms.CharField(required=False)
+    mp_file_types = forms.CharField(max_length=255, required=False)
+    mp_program_type = forms.CharField(max_length=255)
+    # allow user to upload a json schema file
+    mi_json_schema_file = forms.FileField(required=False)
+
+    def clean_version(self):
+        version = self.cleaned_data['version'].strip()
+        return version
+
+    def clean_programming_languages(self):
+        langauge_string = self.cleaned_data['programming_languages']
+        if langauge_string:
+            # generate a list of strings
+            languages = langauge_string.split(',')
+            languages = [lang.strip() for lang in languages]
+            return languages
+        return []
+
+    def clean_operating_systems(self):
+        os_string = self.cleaned_data['operating_systems']
+        if os_string:
+            # generate a list of strings
+            op_systems = os_string.split(',')
+            op_systems = [op.strip() for op in op_systems]
+            return op_systems
+        return []
+
+    def clean_mp_file_types(self):
+        mp_file_types_string = self.cleaned_data['mp_file_types']
+        mp_file_types_dict = dict()
+        if mp_file_types_string:
+            mp_file_types = mp_file_types_string.split(";")
+            for mp_file_type_string in mp_file_types:
+                mp_file_type_lst = mp_file_type_string.split(":")
+                if len(mp_file_type_lst) != 2:
+                    self.add_error("mp_file_types", "Model program file type input data invalid")
```

Comment:
This should be explicit about what is wrong

Reply:
Have made the error message bit better.